proposals/0002-spatial-navigation.md

```diff
@@ -0,0 +1,143 @@
+- Start Date: 2023-09-12
+- PR: https://github.com/videojs/rfcs/pull/2
+
+# Summary
+[summary]: #summary
+
+This feature aims to improve user experience and accessibility on smartTV devices by enabling seamless navigation through interactive elements within the Video.js player using remote control arrow keys.
+
+# Motivation
+[motivation]: #motivation
+
+Smart TV devices commonly use directional input, such as arrow keys on a remote control. By implementing spatial navigation using the spatial navigation polyfill, we aim to ensure that users can intuitively navigate through player controls and elements on smartTV devices.
+
+### Goals
+- Enable seamless and intuitive navigation within the Video.js player on smart TV platforms.
+- Create a new SpatialNavigation component.
+- Implement the spatial navigation polyfill.
+- Redesign the Video.js user interface to suit smart TV platforms and facilitate spatial navigation.
+
+### Non-Goals
+- Keycode mapping for smartTV devices that use non-standardized keycodes. The spatial navigation polyfill uses 37: 'left', 38: 'up', 39: 'right', and 40: 'down'.
+
+# Guide-level Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Smart TV users will be able to navigate through Video.js player controls using arrow keys on their remote controls. The player will have an initial focus; when a user presses, for example, the "right" arrow key, the focus will shift to the next interactive control, allowing them to easily navigate through player UI elements. Focused UI elements will be selectable using the ENTER/OK button on the RCU.
+
+In order to allow the user to navigate out of the progress bar (Slider), the UP and DOWN buttons will not call the stepForward and stepBack functions, as is now. The LEFT and RIGHT buttons will retain the same functions but only when the progress bar (Slider) is in focus.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Dependency Management and Configuration Options
+For this feature, additional dependencies will be added: spatial-navigation-polyfill and flat-polyfill. Flat-polyfill is needed to accommodate devices lacking support for Array.prototype.flat.
+
+To introduce configurational flexibility, a new configuration option `enableRCUNavigation` will be introduced. This option will enable or disable the spatial navigation feature.
+
+## Bundle Size Considerations
+At this stage, we haven't precisely determined the impact of the spatial navigation feature on the bundle size. If the integration of spatial navigation results in a significant growth in the bundle size, we‘ll introduce an alternative Video.js bundle with a spatial navigation feature included.
+
+## Code and UI changes
+
+This new feature will follow these steps:
+- Introduce a public option named `enableRCUNavigation`.
+- When a player is set up with the rcu navigation feature active, initiate the SpatialNavigation instance.
```

Comment:
What is the workflow for users who may already be using Video.js in a Smart TV context?

I'm thinking here that they may have implemented their own spatial navigation handling in their web app. Would they simply leave the option disabled/inactive?

Reply:
I believe they simply keep it disabled (simply do nothing, because it is disabled by default) if they use video.js only for playback and do not use video.js UI or if they've already implemented integration for their app's spatial navigation solution and video.js UI.

Reply:
This new feature will not impact in any way the current Video.js users if they keep it disabled.

Reply:
Added with the latest commit.